### PR TITLE
zkApps state initialisation test

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds tests using a demo zkApp that has an initialisation method and a state update method.

Note that currently the tests do not pass due to bugs in the protocol.

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

File: src/lib/transaction_snark/test/zkapps_examples/initialize_state/initialize_state.ml [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+          let account =
+            Account.create account_id
+              Currency.Balance.(
+                Option.value_exn (add_amount zero (Currency.Amount.of_int 500)))
+          in
+          let _, _ =
+            Ledger.get_or_create_account ledger account_id account
+            |> Or_error.ok_exn
+          in
+          let target_ledger = apply_parties ledger [ parties ] in
```
Comment (deepthiskumar) [inline src/lib/transaction_snark/test/zkapps_examples/initialize_state/initialize_state.ml:208]: use `check_parties_with_merges_exn` instead? It checks that Ledger.apply_transaction matches the what the one in the proof and also does the merge proof of parties

File: src/lib/transaction_snark/test/zkapps_examples/initialize_state/initialize_state.ml [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+      let zkapp_state = (Option.value_exn account.zkapp).app_state in
+      Pickles_types.Vector.iter
+        ~f:(fun x -> assert (Snark_params.Tick.Field.(equal one) x))
+        zkapp_state
+
+    let%test_unit "Update without initialize fails" =
+      let account =
+        test_parties [ Deploy_party.party; Update_state_party.party ]
+      in
+      assert (Option.is_none account.zkapp)
```
Comment (deepthiskumar) [inline src/lib/transaction_snark/test/zkapps_examples/initialize_state/initialize_state.ml:252]: add a case where the zkapp account is new? in which case the account shouldn't be created at all

**Review by deepthiskumar — COMMENTED** [review]
Nice!

File: src/lib/transaction_snark/test/zkapps_examples/initialize_state/initialize_state.ml [hunk 3]
```diff
@@ -240,32 +241,42 @@ let%test_module "Initialize state test" =
...
       in
-      assert (Option.is_none account.zkapp)
+      assert (Option.is_none (Option.value_exn account).zkapp)
+
+    let%test_unit "Initialize without deploy fails" =
+      let account =
+        Option.try_with (fun () ->
+            test_parties [ Initialize_party.party; Update_state_party.party ])
+      in
+      assert (Option.is_none account)
```
Comment (deepthiskumar) [inline src/lib/transaction_snark/test/zkapps_examples/initialize_state/initialize_state.ml:281]: The fee payer and zkapp party are the same account so the account returned here should exist. I'm confused how this test is passing

Comment (mrmr1993) [reply src/lib/transaction_snark/test/zkapps_examples/initialize_state/initialize_state.ml:281]: Applying the transaction fails a check in your harness. It's actually an `_ option option`, where the outer option falls due to the error.

**Review by deepthiskumar — APPROVED** [review]

**Pull request merged**